test/metabase/db/util_test.clj

```diff
@@ -0,0 +1,60 @@
+(ns metabase.db.util-test
+  (:require
+   [clojure.test :refer [deftest testing is]]
+   [metabase.db.util :as mdb.u]
+   [metabase.models.setting :refer [Setting]]
+   [toucan2.core :as t2]))
+
+(set! *warn-on-reflection* true)
+
+(deftest idempotent-insert!-test
+  ;; We test both a case where the database protects against duplicates, and where it does not.
+  ;; Using Setting is perfect because it has only two required fields - (the primary) key & value (with no constraint).
+  ;;
+  ;; In the `:key` case using the `idempotent-insert!` rather than an `or` prevents the from application throwing an
+  ;; exception when there are race conditions. For `:value` it prevents us silently inserting duplicates.
+  ;;
+  ;; It's important to test both, as only the latter has a phantom read issue and thus requires serializable isolation.
+  (let [columns [:key :value]]
+    (doseq [search-col columns]
+      (testing (format "When the search column %s a uniqueness constraint in the db"
+                       (if (= :key search-col) "has" "does not have"))
+
+        ;; We cannot use with-temp, as it starts its own transaction, which stops us setting the isolation level.
+        (let [search-value (str (random-uuid))
+              other-col    (first (remove #{search-col} columns))]
+          (try
+            ;; ensure there is no database detritus
+            (t2/delete! Setting search-col search-value)
+
+            (let [threads  5
+                  promises (atom [])
+                  thunk    (fn []
+                             (mdb.u/idempotent-insert!
+                               (t2/select-one-pk Setting search-col search-value)
```

Comment:
Now that I think of it this test would be stronger and read clearer if you used `t2/select-one` and `t2/insert-returning-instance!`
e.g.
```suggestion
                               (t2/select-one Setting search-col search-value)
```
At the moment the tests could pass even if every thread deleted any existing row and (re)created one.

Reply:
I don't understand how you think we'd get a false pass - the atom stores the id that each thread created, so we'd fail the assertion that they all returned the same value. Am I missing something?

Reply:
Sorry, could have been clearer: the tests could pass only on the `search-col=key` iteration. 

Why? The PK for Setting is the `key` column. When `search-col=key` the `key` returned by `thunk` will be the same for each thread, the random-uuid of line 24. So regardless of whether a new row was inserted or a previously created row was selected, each thread will return the same key. That's not true if you return the whole row.

Reply:
Ah, got you. I think there's some toucan metadata in there that caused issues, but perhaps I can strip that.